nomad/job_endpoint_validators.go

```diff
@@ -0,0 +1,70 @@
+package nomad
+
+import (
+	"strings"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/pkg/errors"
+)
+
+type jobNamespaceConstraintCheckHook struct {
+	srv *Server
+}
+
+func (jobNamespaceConstraintCheckHook) Name() string {
+	return "namespace-constraint-check"
+}
+
+func (c jobNamespaceConstraintCheckHook) Validate(job *structs.Job) (warnings []error, err error) {
+	// This was validated before and matches the WriteRequest namespace
+	ns, err := c.srv.State().NamespaceByName(nil, job.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	if ns == nil {
+		return nil, errors.Errorf("job %q is in nonexistent namespace %q", job.ID, job.Namespace)
+	}
+
+	var disallowedDrivers []string
+	for _, tg := range job.TaskGroups {
+		for _, t := range tg.Tasks {
+			if !taskValidateDriver(t, ns) {
+				disallowedDrivers = append(disallowedDrivers, t.Driver)
+			}
+		}
+	}
+	if len(disallowedDrivers) > 0 {
+		if len(disallowedDrivers) == 1 {
+			return nil, errors.Errorf(
+				"used task driver '%s' is not allowed in namespace %s", disallowedDrivers[0], ns.Name)
```

Comment:
The `%q` formatter quotes this nicely (and consistently with how we do it elsewhere):

```suggestion
				"used task driver %q is not allowed in namespace %s", disallowedDrivers[0], ns.Name)
```